teps/0044-data-locality-and-pod-overhead-in-pipelines.md

```diff
@@ -697,179 +738,73 @@ finally:
+when that solution would address the issues detailed above.
 
-For example:
+### Create a TaskGroup abstraction
 
-```yaml
-kind: TaskGroup
-metadata:
-  name: build-test-deploy
-spec:
- workspaces:
-  - name: source-code
-  - name: test-results
- tasks:
- - name: get-source
-   workspaces:
-   - name: source-code
-     workspace: source-code
-   taskRef:
-     name: git-clone
-   params:
-   - name: url
-      value: $(params.url)
-    - name: revision
-      value: $(params.revision)
- - name: run-unit-tests
-   runAfter: get-source
-   taskRef:
-     name: just-unit-tests
-   workspaces:
-   - name: source-code
-     workspcae: source-code
-   - name: test-results
-     workspace: test-results
- - name: upload-results
-   runAfter: run-unit-tests
-   taskRef:
-     name: gcs-upload
-   params:
-   - name: location
-     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
-   workspaces:
-   - name: data
-     workspace: test-results
-```
+In this approach we create a new Tekton type called a "TaskGroup", which can be implemented as a new CRD or a Custom Task.
+TaskGroups may be embedded in Pipelines. We could create a new TaskGroup controller or use the existing TaskRun controller
+to schedule a TaskGroup.
 
-We could decide if we only support sequential execution, or support an entire DAG. Maybe even finally?
+The controller would be responsible for creating one TaskRun per Task in the TaskGroup, and determining
+at runtime how to schedule each TaskRun. For example, it could create a pod and schedule all the TaskRuns on it,
+or, if a single pod running all the Tasks is too large to be scheduled, it could split the TaskRuns between multiple pods.
+The controller would be responsible for reconciling both the TaskGroup and the TaskRuns created from the TaskGroup.
+We could introduce configuration options to specify whether the controller should attempt to split up TaskRuns
+or simply fail if a single pod wouldn't be schedulable. 
 
-An alternative to the above (which is an ["authoring time"](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability)
-solution) would be a "runtime" CRD, e.g. `TaskGroupRun` which could be passed multiple Tasks and run them together.
+The controller would need to determine how many TaskRuns are needed when the TaskGroup is first reconciled, due to
+[limitations associated with dynamically creating Tasks](#dynamically-created-tasks-in-pipelines).
+When the TaskGroup is first reconciled, it would create all TaskRuns needed, with those that are not ready to execute marked as "pending",
+and a pod with one container per TaskRun. The TaskGroup would store references to any TaskRuns created, and Task results would be stored on the TaskRuns.
 
 Pros:
-* Could be used to define different execution strategies
+* Creating a single TaskRun for each Task would allow individual Task results to be surfaced separately.
+* Allows us to choose which Pipeline features to support.
```

Comment:
I think this is one of the nice pros - since it marks a clear distinction for users.

Reply:
noted in the PR!